Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -291,60 +291,32 @@ public struct DeleteOptions: Encodable {
 public struct InsertOneResult {
     /// The identifier that was inserted. If the document doesn't have an identifier, this value
     /// will be generated and added to the document before insertion.
-    public let insertedId: BsonValue
+    public let insertedId: BsonValue?
 }
 
 /// The result of an `insertMany` command on a `MongoCollection`. 
 public struct InsertManyResult {
     /// Map of the index of the inserted document to the id of the inserted document.
-    public let insertedIds: [Int64: BsonValue]
-
-    /// Given an ordered array of insertedIds, creates a corresponding InsertManyResult.
-    internal init(fromArray arr: [BsonValue]) {
-        var inserted = [Int64: BsonValue]()
-        for (i, id) in arr.enumerated() {
-            let index = Int64(i)
-            inserted[index] = id
-        }
-        self.insertedIds = inserted
-    }
+    public let insertedIds: [BsonValue?]
 }
 
 /// The result of a `delete` command on a `MongoCollection`. 
-public struct DeleteResult {
+public struct DeleteResult: Decodable {
     /// The number of documents that were deleted.
     public let deletedCount: Int
-
-    /// Given a server response to a delete command, creates a corresponding
-    /// `DeleteResult`. If the `from` Document does not have a `deletedCount`
-    /// field, the initialization will fail.
-    internal init?(from: Document) {
-        guard let deletedCount = from["deletedCount"] as? Int else { return nil }
-        self.deletedCount = deletedCount
-    }
 }
 
 /// The result of an `update` operation a `MongoCollection`.
-public struct UpdateResult {
+public struct UpdateResult: Decodable {
     /// The number of documents that matched the filter.
     public let matchedCount: Int
 
     /// The number of documents that were modified.
     public let modifiedCount: Int
 
     /// The identifier of the inserted document if an upsert took place.
-    public let upsertedId: BsonValue?
-
-    /// Given a server response to an update command, creates a corresponding
-    /// `UpdateResult`. If the `from` Document does not have `matchedCount` and
-    /// `modifiedCount` fields, the initialization will fail. The document may
-    /// optionally have an `upsertedId` field.
-    internal init?(from: Document) {
-         guard let matched = from["matchedCount"] as? Int, let modified = from["modifiedCount"] as? Int else {
-            return nil
-         }
-         self.matchedCount = matched
-         self.modifiedCount = modified
-         self.upsertedId = from["upsertedId"]
-    }
+    public let upsertedId: AnyBsonValue?
+
+    /// The number of documents that were upserted.
+    public let upsertedCount: Int
```

Comment:
Does libmongoc actually populate an `upsertedCount` field? I don't see it mentioned in http://mongoc.org/libmongoc/current/mongoc_collection_update_one.html#returns.

PHPC only uses libmongoc's bulk API (see: [WriteResult.c](https://github.com/mongodb/mongo-php-driver/blob/master/src/MongoDB/WriteResult.c)), so we have more fields to work with. In the event that `upsertedCount` isn't available, I think you'll need to infer it yourself based on whether the result document contains an `upsertedId` field. And `upsertedCount` in your result object is necessary for users to differentiate a `null` upserted ID from `upsertedId` not being set (i.e. no upsert occurred).

Reply:
while it is not mentioned in the docs, the reply I get back libmongoc looks like `{ "modifiedCount" : 1, "matchedCount" : 1, "upsertedCount" : 0 }`. 

Jesse's commit message [here](https://github.com/mongodb/mongo-c-driver/commit/84d7dda4232f102fec5b090b7a863df6139fa480) says it was added to mongoc_collection_update_one, update_many, and replace_one.

Reply:
PR opened: https://github.com/mongodb/mongo-c-driver/pull/527